src/dtx/dtx_resync.c

```diff
@@ -73,36 +73,72 @@ dtx_resync_commit(uuid_t po_uuid, struct ds_cont_child *cont,
 	struct dtx_entry		*dte = NULL;
 	int				 rc = 0;
 	int				 i = 0;
+	int				 j = 0;
 
 	D_ASSERT(drh->drh_count >= count);
 
 	D_ALLOC_ARRAY(dte, count);
 	if (dte == NULL)
 		return -DER_NOMEM;
 
-	do {
+	for (i = 0; i < count; i++) {
 		dre = d_list_entry(drh->drh_list.next,
 				   struct dtx_resync_entry, dre_link);
-		dte[i].dte_xid = dre->dre_xid;
-		dte[i].dte_oid = dre->dre_oid;
-
-		rc = vos_dtx_add_cos(cont->sc_hdl, &dre->dre_oid, &dre->dre_xid,
-				     dre->dre_hash, crt_hlc_get(),
-				     dre->dre_intent == DAOS_INTENT_PUNCH ?
-				     true : false);
-		if (rc != 0)
-			D_WARN("Fail to add DTX "DF_DTI" to CoS cache: %d.\n",
+		/* Someone (the DTX owner or batched commit ULT) may have
+		 * committed or aborted the DTX during we handling other
+		 * DTXs. So double check the on-disk status before current
+		 * commit.
+		 */
+		rc = vos_dtx_check(cont->sc_hdl, &dre->dre_xid);
```

Comment:
The DTX resync ULT will batched commit the committable DTXs. There is CPU yield between vos_dtx_check() and former vos_dtx_check() called in dtx_status_handle(). During such internal, some others may commit/abort the DTX.

There are several ULT can commit a DTX:
1) the DTX owner can sync commit it under some cases.
2) the batched commit ULT can do that.
3) DTX resync ULT can do that.

Then which one you want to mange the DTX exclusively?

Reply:
I think it should only be one commit ULT.  if DTX owner or resync needs to commit a few DTX, it can tell the commit ULT, which can always manage to fulfill these requests, which should not be hard IMHO. I think code would be much cleaner and simpler IMHO.

Reply:
I do not think so. The ABT model is not good at the communication between ULTs, that should be avoid as possible as much.

Reply:
what do you mean not good at communication? we can simply communicate by shared structure like self-defined event, it should be easy.

Reply:
Two ULTs cannot run in parallel, if we put something on some shared structure, then we have to yield the CPU and wait the other ULT to execute. I think that is inefficient than each ULT to directly execute related operation if condition is ready.

Reply:
Not sure what you mean here. The yield will happen anyway, no matter what solution. that is why you keep checking the dtx status?

What I mean instead of keep checking dtx status, which makes the whole process quite confused (IMHO), why not only manage dtx in a single ULT. And others just tell them what to do, commit/abort what dtx etc.   I do not think there is much overhead here.

Reply:
We cannot use single ULT, because if we want to use single ULT, then it must the batched commit ULT. But we cannot prevent  the DTX owner to commit/abort the DTX by itself.

Reply:
Not sure I follow, why a single ULT can not do everything about ULT, like commit/abort ?

Reply:
What is your expected single ULT? We already have the DTX batched commit ULT, so it will be the ULT you want, right? But we can NOT prevent the DTX owner (that is related update/punch RPC service ULT) to commit/abort its own DTX, right? So there will always be DTX related race between them.

Reply:
what I mean dtx will be managed by a single ULT,  so dtx check/commit/changing status will always done by this ULT. other ULT can query the dtx status of course.  

There is only one core per stream in anyway, multiple ULT to operate DTX will not provide any benefits, just troublesome IMHO.

Reply:
It is impossible as make single ULT to manage the DTXs, because of each of RPC service ULT can create/commit/abort the its own DTX.

Reply:
of course echo I/O should create/manage its own dtx before getting reply from all replicas, resync will not care about these DTX at all.

But after that, the I/O ULT should hand these DTX to this single ULT. then all batched commit/abort/resync should be done by this single ULT. We can avoid a lot complexity by doing this.

Reply:
> of course echo I/O should create/manage its own dtx before getting reply from all replicas, resync will not care about these DTX at all.
> 
How the DTX resync logic can know whether the DTX owner is still alive or not?

> But after that, the I/O ULT should hand these DTX to this single ULT. then all batched commit/abort/resync should be done by this single ULT. We can avoid a lot complexity by doing this.

Batched commit ULT and DTX resync ULT has each own duty, it is not impossible to merge their tasks, but such ULT still need to handle the race with DTX owner ULT. Then the logic will be more complex.

Reply:
I assume you mean dtx owner ULT is the normal I/O ULT?  If we have this single DTX management ULT, then

1. If there are no leader change, DTX resync process does not have to deal with those alive DTX (uncommittable).  So it simply get the DTX from committable list and hand it over to this DTX single ULT to tell it commit them.

2. If there is leader change, then DTX resync process somehow get those dtx from active list, then also hand them to DTX single ULT.

3. For normal I/O process, after it get the replied from all replicas, it will also hand the DTX to the single ULT.

I.e. always let the single ULT manage the DTX commit process.  I think this will avoid a lot complexity here.

Reply:
In theory, it may work, but from the implementation and efficiency perspective, it is almost unacceptable. For example, if the normal I/O process found that it cannot complete some modification and has to abort the DTX, according to current implementation, it will abort the DTX on both local and remote replicas via dtx_abort(), and then reply to the client. But if giving such control to some dedicated ULT, then the normal I/O ULT will add such DTX into some list, and then notify such dedicated ULT to handle such DTX, then wait on some ABT conditions until such DTX is aborted. If the system very idle, it is not problem; but if such dedicated ULT has a lot of DTX to handle, related normal I/O ULTs to may have to wait some long time. Different from the commit case, for most of time, commit can be async mode, but abort is sync operation. Although abort case is rare case than commit, but it is more frequent than rebuild case, especially under the conditional update.

On the other hand, the DTX is created by the normal I/O ULT, it is the owner, allowing the owner to synchronously commit/abort such DTX is the most intuitive and most efficient way. The dedicated DTX ULT is only for asynchronous batched commit for performance consideration. But in your model, the perfjoamcne/efficiency will be worse, that is totally deviate from the purpose.

In fact, the total lines for the dtx_resync_commit() and dtx_status_handle() are abort 200 LOC, that includes the a lot of detailed comments, variables definition and empty lines. So the core logic for that is just ten of LOC. So I do not think it is real so complex, are you sure that it will be better with dedicated DTX? Here may be less, but other places will be more.

Reply:
Hmm, if it is the DTX needs to be aborted or synchronized, it can be handled by IO ULT, and no one else will care its status anyway.  I.e. both resync or batched commit will not change this dtx status.  And I suggest the following 

1. So synch/abort DTX should be handle by IO ULT.
2. Other DTXs(batched commit or resync for rebuild) should be handled by this single ULT.  

The problem I have with the current implementation is that one DTX might be handled by different ULTs, but without notifying each others.  So it caused the DTX resync keeping checking is status, which easily causing confusion and bugs IMHO.  So I would suggest during dtx resync should only check 

1. the committable DTX (if leader is not changed).
2. And also iterate the DTX from active list, but only handle those DTX if it is leader, but no leader flags in its entry, i.e. new leader.

Then hand these DTXs to this single DTX ULT to commit them,  and during this process no one else will change its status at all.

Reply:
> . And I suggest the following
> 
> 1. So synch/abort DTX should be handle by IO ULT.

That is current implementation.

> 2. Other DTXs(batched commit or resync for rebuild) should be handled by this single ULT.

That helps nothing!
Please consider the situation that some DTX was not completed before the old leader crashed. Then on the new leader, the DTX-resync ULT will check its status with other available replicas. If it is committable, as your suggested, the DTX-resync ULT gives it to the batched commit ULT. But at that time, related client may be still alive, it resends related RPC to the new leader, that may also commit its DTX, then there is race between the normal I/O ULT and the batched commit ULT, the same situation as current patch does. If such DTX is not committable, the DTX-resync ULT will want to abort such DTX, then as you suggested, the DTX-resync ULT gives it to the batched commit ULT. But before the batched commit ULT aborting such DTX, the client may resend the RPC to the new leader and re-excute successfully, then it become committable. So when the batched commit ULT wants to abort such DTX, it still needs to do the similar check as current patch does!

On the other hand, there is only single DTX batched commit ULT per VOS target. If you give many tasks to such ULT, it may cannot handle them in time, as to there are more chances to race with other normal I/O ULTs.

> 
> The problem I have with the current implementation is that one DTX might be handled by different ULTs, but without notifying each others. So it caused the DTX resync keeping checking is status, which easily causing confusion and bugs IMHO. So I would suggest during dtx resync should only check
> 
> 1. the committable DTX (if leader is not changed)
> 2. And also iterate the DTX from active list, but only handle those DTX if it is leader, but no leader flags in its entry, i.e. new leader.
> 
> Then hand these DTXs to this single DTX ULT to commit them, and during this process no one else will change its status at all.

Please note there is resent RPC.

Reply:
> > . And I suggest the following
> > 
> > 1. So synch/abort DTX should be handle by IO ULT.
> 
> That is current implementation.
> 
> > 1. Other DTXs(batched commit or resync for rebuild) should be handled by this single ULT.
> 
> That helps nothing!
> Please consider the situation that some DTX was not completed before the old leader crashed. Then on the new leader, the DTX-resync ULT will check its status with other available replicas. If it is committable, as your suggested, the DTX-resync ULT gives it to the batched commit ULT. But at that time, related client may be still alive, it resends related RPC to the new leader, that may also commit its DTX, then there is race between the normal I/O ULT and the batched commit ULT, the same situation as current patch does. If such DTX is not committable, the DTX-resync ULT will want to abort such DTX, then as you suggested, the DTX-resync ULT gives it to the batched commit ULT. But before the batched commit ULT aborting such DTX, the client may resend the RPC to the new leader and re-excute successfully, then it become committable. So when the batched commit ULT wants to abort such DTX, it still needs to do the similar check as current patch does!
> 
> On the other hand, there is only single DTX batched commit ULT per VOS target. If you give many tasks to such ULT, it may cannot handle them in time, as to there are more chances to race with other normal I/O ULTs.
> 
> > The problem I have with the current implementation is that one DTX might be handled by different ULTs, but without notifying each others. So it caused the DTX resync keeping checking is status, which easily causing confusion and bugs IMHO. So I would suggest during dtx resync should only check
> > 
> > 1. the committable DTX (if leader is not changed)
> > 2. And also iterate the DTX from active list, but only handle those DTX if it is leader, but no leader flags in its entry, i.e. new leader.
> > 
> > Then hand these DTXs to this single DTX ULT to commit them, and during this process no one else will change its status at all.
> 
> Please note there is resent RPC.

Hmm, I think resend case can be resolved, there are a few options.

1. Using different DTX,  i.e. if the client did not get the reply from the original leader, and then send the request to the new leader, then it can use the new DTX anyway IMHO, since the old DTX can be aborted by the resync process anyway.

2. If we choose to use the same DTX ( I prefer not, do not see much benefits here), then we  can serialize the resync DTX (only the dtx with new leader) and resend object process.  i.e. server resync these DTX (changing leader), then can process those resend request to the new leader.  This process might be able to optimize a bit.

Personally I prefer 1,  unless we have to use the same DTX.

Reply:
> > Please note there is resent RPC.
> 
> Hmm, I think resend case can be resolved, there are a few options.
> 
> 1. Using different DTX,  i.e. if the client did not get the reply from the original leader, and then send the request to the new leader, then it can use the new DTX anyway IMHO, since the old DTX can be aborted by the resync process anyway.

We have to use the same DTX, because the former RPC may have executed successfully on all replicas and the DTX was committable, but the old leader crashed before replying to the client. If we use different DTX, the result will be wrong.

> 2. If we choose to use the same DTX ( I prefer not, do not see much benefits here), then we  can serialize the resync DTX (only the dtx with new leader) and resend object process.  i.e. server resync these DTX (changing leader), then can process those resend request to the new leader.  This process might be able to optimize a bit

It is not good. Currently, DAOS pool version changing will trigger object rebuild, object rebuild will trigger DTX resync. But the pool map version changing is not per redundancy group based. Mean that even if the failed target is nothing related with current redundancy group, we still run DTX resync for current redundancy group. If we block resent RPC under such case, it is too much overhead.
.
> 
> Personally I prefer 1, unless we have to use the same DTX.

Reply:
> > > Please note there is resent RPC.
> > 
> > 
> > Hmm, I think resend case can be resolved, there are a few options.
> > 
> > 1. Using different DTX,  i.e. if the client did not get the reply from the original leader, and then send the request to the new leader, then it can use the new DTX anyway IMHO, since the old DTX can be aborted by the resync process anyway.
> 
> We have to use the same DTX, because the former RPC may have executed successfully on all replicas and the DTX was committable, but the old leader crashed before replying to the client. If we use different DTX, the result will be wrong.

Hmm, if the old leader already reply to client, then the new leader has to resync and commit the transaction.  Otherwise it can abort the transaction. But the problem is the new leader does not know whether client get reply.  Indeed, so it has to use the same DTX here.

> 
> > 1. If we choose to use the same DTX ( I prefer not, do not see much benefits here), then we  can serialize the resync DTX (only the dtx with new leader) and resend object process.  i.e. server resync these DTX (changing leader), then can process those resend request to the new leader.  This process might be able to optimize a bit
> 
> It is not good. Currently, DAOS pool version changing will trigger object rebuild, object rebuild will trigger DTX resync. But the pool map version changing is not per redundancy group based. Mean that even if the failed target is nothing related with current redundancy group, we still run DTX resync for current redundancy group. If we block resent RPC under such case, it is too much overhead.


If the failed target is nothing related with the current redundancy group, it will not resend the request. My suggestion is only delay handling resend RPC until the new leader build its dtx list.  So during resync, each xstream just build these DTX list for changing leader case, then if the leader receive a resend RPC from client, it will check the list to see if resyc process will handle it.  Hmm, we probably can go even further here, i.e. each non-leader can cache the such DTX in DRAM.  i.e. when the non-leader get any update request from leader, it should cache such DTX status in DRAM, there might be a lot benefits, Jeff and I even discussed a bit about this today.  Anyway I think this is a good topic, probably worth a broader attention here.



















> .
> 
> > Personally I prefer 1, unless we have to use the same DTX.